File: .idea/app/src/main/java/com/chevstrap/rbx/UI/Elements/Settings/Pages/ChevstrapFragment.java
package com.chevstrap.rbx.UI.Elements.Settings.Pages;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.chevstrap.rbx.App;
import com.chevstrap.rbx.ConfigManager;
import com.chevstrap.rbx.Extensions.CustomUIComponents;
import com.chevstrap.rbx.Models.MethodPair;
import com.chevstrap.rbx.R;
import com.chevstrap.rbx.SettingsActivity;

import org.json.JSONArray;

import java.util.HashMap;
import java.util.Objects;

public class ChevstrapFragment extends Fragment {
	private final HashMap<Integer, LinearLayout> layoutMap = new HashMap<>();
	ConfigManager manager;
	private LinearLayout linear1;
	private LinearLayout linear2;
	private ActivityResultLauncher<String> videoPickerLauncher;
	private TextView textview_description;

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.a_normal_page_fragment, container, false);
		manager = ConfigManager.getInstance();

		try {
			initialize(savedInstanceState, view);
			initializeLogic();
		} catch (Exception e) {
			App.getLogger().writeException("ChevstrapFragment::initialize", e);
		}

		videoPickerLauncher = registerForActivityResult(
				new ActivityResultContracts.GetContent(),
				uri -> {
					if (uri == null) {
						App.getLogger().writeLine("ChevstrapFragment::PersistUriPermission", "No video selected");
						return;
					}

					try {
						requireContext().getContentResolver().takePersistableUriPermission(
								uri,
								Intent.FLAG_GRANT_READ_URI_PERMISSION
						);
						App.getLogger().writeLine("ChevstrapFragment::PersistUriPermission", "Persisted URI permission: " + uri);
					} catch (SecurityException e) {
						App.getLogger().writeException("ChevstrapFragment::PersistUriPermission", e);
					}

					App.getChevstrapSettings().setDisplayVideoThemeWithUri(uri.toString());
					App.getLogger().writeLine("ChevstrapFragment::PersistUriPermission", "Saved video URI: " + uri);

					Activity settingsActivity = App.getSavedSettingsActivity();
					if (settingsActivity instanceof SettingsActivity) {
						((SettingsActivity) settingsActivity).playVideo(uri);
						App.getLogger().writeLine("ChevstrapFragment::PersistUriPermission", "Playing video on SettingsActivity");
					} else {
						App.getLogger().writeLine("ChevstrapFragment::PersistUriPermission", "Settings activity not found");
					}
				}
		);

		App.setSavedFragmentActivity(this.requireActivity());
		return view;
	}

	@Override
	public void onDestroyView() {
		super.onDestroyView();

		for (LinearLayout layout : layoutMap.values()) {
			if (layout != null) {
				layout.removeAllViews();
			}
		}
		layoutMap.clear();
		linear2 = null;
		manager = null;
	}

	private void initialize(Bundle savedInstanceState, View view) {
		linear2 = view.findViewById(R.id.linear2);
		linear1 = view.findViewById(R.id.linear1);
		textview_description = view.findViewById(R.id.textview_description);

		boolean isDark = Objects.equals(App.getChevstrapSettings().getAppThemeInApp(), "dark");
		if (isDark) {
			textview_description.setTextColor(Color.parseColor("#D6D6D6"));
		} else {
			textview_description.setTextColor(Color.parseColor("#282828"));
		}
	}

	private void initializeLogic() {
		AstyleButton1(linear1);
	}

	public void addSection(ViewGroup parent, String text) {
		CustomUIComponents.SectionResult section = CustomUIComponents.addSection(getContext(), text);
		parent.addView(section.sectionContainer);
	}

	public void addDivider(ViewGroup parent) {
		CustomUIComponents.DividerResult divider = CustomUIComponents.addDivider(getContext());
		parent.addView(divider.dividerView);
	}

	public void addToggle(String name, String description, MethodPair selectedMethod, int indexParented) {
		CustomUIComponents.ToggleResult getCustomToggle = CustomUIComponents.addToggle(getContext(), name, description, linear2, selectedMethod);
		if (indexParented > -1) {
			Objects.requireNonNull(layoutMap.get(indexParented)).addView(getCustomToggle.toggleView);
		} else {
			linear2.addView(getCustomToggle.toggleView);
		}
	}

	public void addImportButton(String name, String description, String command, String typeCommand, int indexParented) {
		CustomUIComponents.ButtonResult getCustomButton = CustomUIComponents.addImportButton(getContext(), name, description, linear2);
		if (indexParented > -1) {
			Objects.requireNonNull(layoutMap.get(indexParented)).addView(getCustomButton.buttonView);
		} else {
			linear2.addView(getCustomButton.buttonView);
		}

		getCustomButton.buttonOne.setOnClickListener(v -> {
			if ("set_animated_theme".equals(typeCommand)) {
				videoPickerLauncher.launch("video/*");
			}
		});
	}

	public void addAccordionMenu(String name, String description, Runnable AFunction, int index) {
		CustomUIComponents.AccordionMenuResult getCustomButton =
				CustomUIComponents.addAccordionMenu(getContext(), name, description, linear2, AFunction);

		layoutMap.put(index, getCustomButton.expandContainer);
		linear2.addView(getCustomButton.buttonView);
	}

	public void addTextbox(String name, String description, MethodPair selectedMethod, String TypeIs, int indexParented) {
		CustomUIComponents.TextboxResult getCustomTextbox = CustomUIComponents.addTextbox(getContext(), name, description, linear2, selectedMethod, TypeIs);
		if (indexParented > -1) {
			Objects.requireNonNull(layoutMap.get(indexParented)).addView(getCustomTextbox.textboxView);
		} else {
			linear2.addView(getCustomTextbox.textboxView);
		}
	}

	public void addDropdown(String name, String description, final JSONArray jsonArray, MethodPair selectedMethod, int indexParented) {
		CustomUIComponents.DropdownResult getCustomDropdown = CustomUIComponents.addDropdown(getContext(), name, description, linear2, jsonArray, selectedMethod);
		if (indexParented > -1) {
			Objects.requireNonNull(layoutMap.get(indexParented)).addView(getCustomDropdown.dropDownView);
		} else {
			linear2.addView(getCustomDropdown.dropDownView);
		}
	}

	@Override
	public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
		LinearLayout parentLayout = view.findViewById(R.id.linear2);
		textview_description.setText(App.getTextLocale(App.getAppContext(), R.string.menu_settings_description));
		try {
			Chevstrap.addEveryPresets(requireContext(), parentLayout, this);
		} catch (Exception e) {
			App.getLogger().writeException("ChevstrapFragment::onViewCreated", e);
		}
	}

	private void AstyleButton1(LinearLayout button) {
		GradientDrawable drawable = new GradientDrawable();
		drawable.setCornerRadius(30);
		drawable.setColor(Color.parseColor("#151515"));
		button.setBackground(drawable);
	}

    public void setLinear1(LinearLayout linear1) {
        this.linear1 = linear1;
    }
}

File: app/src/main/java/com/chevstrap/rbx/ui/views/settings/pages/ChevstrapFragment.kt
package com.chevstrap.rbx.ui.views.settings.pages

import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.graphics.toColorInt
import androidx.fragment.app.Fragment
import com.chevstrap.rbx.App
import com.chevstrap.rbx.ConfigManager
import com.chevstrap.rbx.R
import com.chevstrap.rbx.extensions.ResourceManagerEx
import com.chevstrap.rbx.ui.components.ComponentUtils
import com.chevstrap.rbx.ui.viewModels.MethodAction
import com.chevstrap.rbx.ui.viewModels.MethodPair
import org.json.JSONArray

class ChevstrapFragment : Fragment() {
    private val layoutMap = HashMap<Int, LinearLayout>()
    private var manager: ConfigManager? = null
    private var linear1: LinearLayout? = null
    private var linear2: LinearLayout? = null
    private var textviewDesc: TextView? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view = inflater.inflate(
            R.layout.page_fragment,
            container,
            false
        )

        initialize(view)

        return view
    }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?
    ) {
        super.onViewCreated(view, savedInstanceState)

        try {
            manager = ConfigManager.instance

            initializeLogic()

            val parentLayout = requireNotNull(linear2)

            Chevstrap.addEveryPresets(
                requireContext(),
                parentLayout,
                this
            )

            App.savedFragmentActivity = requireActivity()
        } catch (e: Exception) {
            App.logger.writeException(
                "ChevstrapFragment::onViewCreated",
                e
            )
        }
    }

    override fun onDestroyView() {
        layoutMap.values.forEach {
            it.removeAllViews()
        }

        layoutMap.clear()
        linear1 = null
        linear2 = null
        textviewDesc = null

        super.onDestroyView()
    }

    private fun initialize(view: View) {
        linear1 = view.findViewById(R.id.linear1)
        linear2 = view.findViewById(R.id.linear2)
        textviewDesc = view.findViewById(R.id.textview_desc)
    }

    private fun initializeLogic() {
        textviewDesc?.text = ResourceManagerEx.getStringOrEmpty(
            requireContext(),
            R.string.menu_settings_description
        )

        linear1?.let {
            aStyleButton1(it)
        }
    }

    fun addSection(parent: ViewGroup, text: String?) {
        val section =  ComponentUtils.addSection(context, text)
        parent.addView(section.sectionContainer)
    }

    fun addDivider(parent: ViewGroup) {
        val divider =  ComponentUtils.addDivider(context)
        parent.addView(divider.dividerView)
    }

    fun addToggle(
        name: String?,
        description: String?,
        method: MethodPair<*, Boolean>?,
        indexParented: Int
    ) {
        method ?: return

        val parent = if (indexParented > -1) {
            requireNotNull(layoutMap[indexParented])
        } else {
            requireNotNull(linear2)
        }

        val toggle =  ComponentUtils.addToggle(
            context,
            name,
            description,
            method
        )

        parent.addView(toggle.toggleView)
    }

    fun addButton(
        name: String?,
        description: String?,
        method: MethodAction,
        indexParented: Int
    ) {
        val parent = if (indexParented > -1) {
            requireNotNull(layoutMap[indexParented])
        } else {
            requireNotNull(linear2)
        }

        val button =  ComponentUtils.addButton(
            context,
            name,
            description,

            method
        )

        parent.addView(button.buttonView)
    }

    fun addAccordionMenu(
        name: String?,
        description: String?,
        function: Runnable?,
        index: Int
    ) {
        val parent = requireNotNull(linear2)

        val accordion =  ComponentUtils.addAccordionMenu(
            context,
            name,
            description,
            function
        )

        layoutMap[index] = accordion.expandContainer as LinearLayout
        parent.addView(accordion.buttonView)
    }

    fun addTextbox(
        name: String?,
        description: String?,
        method: MethodPair<*, String>?,
        typeIs: String?,
        indexParented: Int
    ) {
        method ?: return

        val parent = if (indexParented > -1) {
            requireNotNull(layoutMap[indexParented])
        } else {
            requireNotNull(linear2)
        }

        val textbox =  ComponentUtils.addTextbox(
            context,
            name,
            description,
            method,
            typeIs
        )

        parent.addView(textbox.textboxView)
    }

    fun addDropdown(
        name: String?,
        description: String?,
        jsonArray: JSONArray?,
        method: MethodPair<*, String>?,
        indexParented: Int
    ) {
        method ?: return

        val parent = if (indexParented > -1) {
            requireNotNull(layoutMap[indexParented])
        } else {
            requireNotNull(linear2)
        }

        val dropdown =  ComponentUtils.addDropdown(
            context,
            name,
            description,
            jsonArray,
            method
        )

        parent.addView(dropdown.dropDownView)
    }

    private fun aStyleButton1(button: LinearLayout) {
        val drawable = GradientDrawable().apply {
            cornerRadius = 30f
            setColor("#151515".toColorInt())
        }

        button.background = drawable
    }
}